src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs

```diff
@@ -27,90 +29,102 @@ protected SwitchBinder(Binder next, SwitchStatementSyntax switchSyntax)
+                    var boundSwitchExpression = BindSwitchExpression(_switchSyntax.Expression, switchGoverningDiagnostics);
+                    _switchGoverningDiagnostics = switchGoverningDiagnostics;
+                    Interlocked.CompareExchange(ref _switchGoverningExpression, boundSwitchExpression, null);
                 }
+
+                return _switchGoverningExpression;
             }
+        }
+
+        protected TypeSymbol SwitchGoverningType => SwitchGoverningExpression.Type;
 
-            return false;
+        protected DiagnosticBag SwitchGoverningDiagnostics
+        {
+            get
+            {
+                var discarded = SwitchGoverningExpression;
+                return _switchGoverningDiagnostics;
+            }
         }
 
         // Dictionary for the switch case/default labels.
         // Case labels with a non-null constant value are indexed on their ConstantValue.
         // Default label(s) are indexed on a special DefaultKey object.
         // Invalid case labels with null constant value are indexed on the labelName.
-        private Dictionary<object, List<SourceLabelSymbol>> _lazySwitchLabelsMap;
+        private Dictionary<object, SourceLabelSymbol> _lazySwitchLabelsMap;
         private static readonly object s_defaultKey = new object();
 
-        private Dictionary<object, List<SourceLabelSymbol>> SwitchLabelsMap
+        private Dictionary<object, SourceLabelSymbol> LabelsByValue
         {
             get
             {
                 if (_lazySwitchLabelsMap == null && this.Labels.Length > 0)
                 {
-                    _lazySwitchLabelsMap = BuildMap(this.Labels);
+                    _lazySwitchLabelsMap = BuildLabelsByValue(this.Labels);
                 }
 
                 return _lazySwitchLabelsMap;
             }
         }
 
-        private static Dictionary<object, List<SourceLabelSymbol>> BuildMap(ImmutableArray<LabelSymbol> labels)
+        private static Dictionary<object, SourceLabelSymbol> BuildLabelsByValue(ImmutableArray<LabelSymbol> labels)
         {
             Debug.Assert(labels.Length > 0);
 
-            var map = new Dictionary<object, List<SourceLabelSymbol>>(labels.Length, new SwitchConstantValueHelper.SwitchLabelsComparer());
+            var map = new Dictionary<object, SourceLabelSymbol>(labels.Length, new SwitchConstantValueHelper.SwitchLabelsComparer());
             foreach (SourceLabelSymbol label in labels)
             {
                 SyntaxKind labelKind = label.IdentifierNodeOrToken.Kind();
+                if (labelKind == SyntaxKind.IdentifierToken) continue;
 
-                if (labelKind == SyntaxKind.CaseSwitchLabel ||
-                    labelKind == SyntaxKind.DefaultSwitchLabel)
+                object key;
+                var constantValue = label.SwitchCaseLabelConstant;
+                if (constantValue != (object)null && !constantValue.IsBad)
```

Comment:
I'm not sure I understand the question. Are you saying it is preferable to put the cast on the left rather than on the right?

---
In reply to: [70298683](https://github.com/dotnet/roslyn/pull/12416#discussion_r70298683) [](ancestors = 70298683)